File: FruitWholesale.Client/src/app/features/reports/report.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { environment } from '../../../environments/environment';
import {
  DailyCollectionReportRow,
  DailyExpenseReportRow,
  DailySalesReportRow,
  ExpenseByCategoryReportRow,
  FruitSalesReportRow,
  OutstandingReportRow,
  ProfitSummaryReportRow,
  PurchaseReportRow,
  SalaryByEmployeeReportRow
} from '../../core/models/report.model';
import { toHttpParams } from '../../core/utils/http-params.util';

@Injectable({ providedIn: 'root' })
export class ReportService {
  private readonly baseUrl = `${environment.apiUrl}/report`;

  constructor(private readonly http: HttpClient) {}

  getDailySales(fromDate: string, toDate: string, shopId?: number | null) {
    return this.http.get<DailySalesReportRow[]>(`${this.baseUrl}/daily-sales`, { params: toHttpParams({ fromDate, toDate, shopId }) });
  }

  getDailyCollection(fromDate: string, toDate: string, shopId?: number | null) {
    return this.http.get<DailyCollectionReportRow[]>(`${this.baseUrl}/daily-collection`, {
      params: toHttpParams({ fromDate, toDate, shopId })
    });
  }

  getDailyExpense(fromDate: string, toDate: string, expenseCategoryId?: number | null) {
    return this.http.get<DailyExpenseReportRow[]>(`${this.baseUrl}/daily-expense`, {
      params: toHttpParams({ fromDate, toDate, expenseCategoryId })
    });
  }

  getPurchaseReport(fromDate: string, toDate: string, supplierId?: number | null) {
    return this.http.get<PurchaseReportRow[]>(`${this.baseUrl}/purchase`, { params: toHttpParams({ fromDate, toDate, supplierId }) });
  }

  getFruitSales(fromDate: string, toDate: string) {
    return this.http.get<FruitSalesReportRow[]>(`${this.baseUrl}/fruit-sales`, { params: toHttpParams({ fromDate, toDate }) });
  }

  getOutstanding() {
    return this.http.get<OutstandingReportRow[]>(`${this.baseUrl}/outstanding`);
  }

  getProfitSummary(fromDate: string, toDate: string) {
    return this.http.get<ProfitSummaryReportRow[]>(`${this.baseUrl}/profit-summary`, { params: toHttpParams({ fromDate, toDate }) });
  }

  getExpenseByCategory(fromDate: string, toDate: string) {
    return this.http.get<ExpenseByCategoryReportRow[]>(`${this.baseUrl}/expense-by-category`, { params: toHttpParams({ fromDate, toDate }) });
  }

  getSalaryByEmployee(fromDate: string, toDate: string, employeeId?: number | null) {
    return this.http.get<SalaryByEmployeeReportRow[]>(`${this.baseUrl}/salary-by-employee`, {
      params: toHttpParams({ fromDate, toDate, employeeId })
    });
  }
}

File: src/FruitWholesale.Application/Services/ReportService.cs
using FruitWholesale.Application.Common.Interfaces;
using FruitWholesale.Application.DTOs.Reports;

namespace FruitWholesale.Application.Services;

public interface IReportService
{
    Task<List<DailySalesReportRow>> GetDailySalesAsync(DateTime fromDate, DateTime toDate, int? shopId);
    Task<List<DailyCollectionReportRow>> GetDailyCollectionAsync(DateTime fromDate, DateTime toDate, int? shopId);
    Task<List<DailyExpenseReportRow>> GetDailyExpenseAsync(DateTime fromDate, DateTime toDate, int? expenseCategoryId);
    Task<List<PurchaseReportRow>> GetPurchaseReportAsync(DateTime fromDate, DateTime toDate, int? supplierId);
    Task<List<FruitSalesReportRow>> GetFruitSalesReportAsync(DateTime fromDate, DateTime toDate);
    Task<List<OutstandingReportRow>> GetOutstandingReportAsync();
    Task<List<ProfitSummaryReportRow>> GetProfitSummaryAsync(DateTime fromDate, DateTime toDate);
    Task<List<ExpenseByCategoryReportRow>> GetExpenseByCategoryAsync(DateTime fromDate, DateTime toDate);
    Task<List<SalaryByEmployeeReportRow>> GetSalaryByEmployeeAsync(DateTime fromDate, DateTime toDate, int? employeeId);
}

public class ReportService(IReportRepository repository) : IReportService
{
    public Task<List<DailySalesReportRow>> GetDailySalesAsync(DateTime fromDate, DateTime toDate, int? shopId) =>
        repository.GetDailySalesAsync(fromDate, toDate, shopId);

    public Task<List<DailyCollectionReportRow>> GetDailyCollectionAsync(DateTime fromDate, DateTime toDate, int? shopId) =>
        repository.GetDailyCollectionAsync(fromDate, toDate, shopId);

    public Task<List<DailyExpenseReportRow>> GetDailyExpenseAsync(DateTime fromDate, DateTime toDate, int? expenseCategoryId) =>
        repository.GetDailyExpenseAsync(fromDate, toDate, expenseCategoryId);

    public Task<List<PurchaseReportRow>> GetPurchaseReportAsync(DateTime fromDate, DateTime toDate, int? supplierId) =>
        repository.GetPurchaseReportAsync(fromDate, toDate, supplierId);

    public Task<List<FruitSalesReportRow>> GetFruitSalesReportAsync(DateTime fromDate, DateTime toDate) =>
        repository.GetFruitSalesReportAsync(fromDate, toDate);

    public Task<List<OutstandingReportRow>> GetOutstandingReportAsync() => repository.GetOutstandingReportAsync();

    public Task<List<ProfitSummaryReportRow>> GetProfitSummaryAsync(DateTime fromDate, DateTime toDate) =>
        repository.GetProfitSummaryAsync(fromDate, toDate);

    public Task<List<ExpenseByCategoryReportRow>> GetExpenseByCategoryAsync(DateTime fromDate, DateTime toDate) =>
        repository.GetExpenseByCategoryAsync(fromDate, toDate);

    public Task<List<SalaryByEmployeeReportRow>> GetSalaryByEmployeeAsync(DateTime fromDate, DateTime toDate, int? employeeId) =>
        repository.GetSalaryByEmployeeAsync(fromDate, toDate, employeeId);
}
